resilience4j-spring/src/main/java/io/github/resilience4j/recovery/RecoveryMethod.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2019 Kyuhyen Hwang
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.recovery;
+
+import io.github.resilience4j.core.lang.Nullable;
+import org.springframework.util.ConcurrentReferenceHashMap;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Reflection utility for invoking recovery method. Recovery method should have same return type and parameter types of original method but the last additional parameter.
+ * The last additional parameter should be a subclass of {@link Throwable}. When {@link RecoveryMethod#recover(Throwable)} is invoked, {@link Throwable} will be passed to that last parameter.
+ * If there are multiple recovery method, one of the methods that has most closest superclass parameter of thrown object will be invoked.
+ * <pre>
+ * For example, there are two recovery methods
+ * {@code
+ * String recovery(String parameter, RuntimeException exception)
+ * String recovery(String parameter, IllegalArgumentException exception)
+ * }
+ * and if try to recover from {@link NumberFormatException}, {@code String recovery(String parameter, IllegalArgumentException exception)} will be invoked.
+ * </pre>
+ */
+public class RecoveryMethod {
+    private static final Map<MethodMeta, Map<Class<?>, Method>> RECOVERY_METHODS_CACHE = new ConcurrentReferenceHashMap<>();
```

Comment:
may I ask why ConcurrentReferenceHashMap ? FYI it is not 100% thread safe as i remember !

Reply:
I found `ConcurrentReferenceHashMap` is used as a cache in [org.springframework.util.ReflectionUtils](
https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/util/ReflectionUtils.java). And I thought cached values could be removed if there is no memory space, so I chose `SoftReference`. Since there is no such warning in the document, I didn't know `ConcurrentReferenceHashMap ` is not 100% thread-safe. But should it be a thread-safe guaranteed? Recovery methods are not changing in run-time.